commons-math/src/main/java/org/apache/commons/math4/fitting/leastsquares/LevenbergMarquardtOptimizer.java

```diff
@@ -0,0 +1,1043 @@
+     */
+    public LevenbergMarquardtOptimizer withParameterRelativeTolerance(double newParRelativeTolerance) {
+        return new LevenbergMarquardtOptimizer(
+                initialStepBoundFactor,
+                costRelativeTolerance,
+                newParRelativeTolerance,
+                orthoTolerance,
+                qrRankingThreshold);
+    }
+
+    /**
+     * Modifies the given parameter.
+     *
+     * @param newOrthoTolerance Desired max cosine on the orthogonality between
+     * the function vector and the columns of the Jacobian.
+     * @return a new instance.
+     */
+    public LevenbergMarquardtOptimizer withOrthoTolerance(double newOrthoTolerance) {
+        return new LevenbergMarquardtOptimizer(
+                initialStepBoundFactor,
+                costRelativeTolerance,
+                parRelativeTolerance,
+                newOrthoTolerance,
+                qrRankingThreshold);
+    }
+
+    /**
+     * @param newQRRankingThreshold Desired threshold for QR ranking.
+     * If the squared norm of a column vector is smaller or equal to this
+     * threshold during QR decomposition, it is considered to be a zero vector
+     * and hence the rank of the matrix is reduced.
+     * @return a new instance.
+     */
+    public LevenbergMarquardtOptimizer withRankingThreshold(double newQRRankingThreshold) {
+        return new LevenbergMarquardtOptimizer(
+                initialStepBoundFactor,
+                costRelativeTolerance,
+                parRelativeTolerance,
+                orthoTolerance,
+                newQRRankingThreshold);
+    }
+
+    /**
+     * Gets the value of a tuning parameter.
+     * @see #withInitialStepBoundFactor(double)
+     *
+     * @return the parameter's value.
+     */
+    public double getInitialStepBoundFactor() {
+        return initialStepBoundFactor;
+    }
+
+    /**
+     * Gets the value of a tuning parameter.
+     * @see #withCostRelativeTolerance(double)
+     *
+     * @return the parameter's value.
+     */
+    public double getCostRelativeTolerance() {
+        return costRelativeTolerance;
+    }
+
+    /**
+     * Gets the value of a tuning parameter.
+     * @see #withParameterRelativeTolerance(double)
+     *
+     * @return the parameter's value.
+     */
+    public double getParameterRelativeTolerance() {
+        return parRelativeTolerance;
+    }
+
+    /**
+     * Gets the value of a tuning parameter.
+     * @see #withOrthoTolerance(double)
+     *
+     * @return the parameter's value.
+     */
+    public double getOrthoTolerance() {
+        return orthoTolerance;
+    }
+
+    /**
+     * Gets the value of a tuning parameter.
+     * @see #withRankingThreshold(double)
+     *
+     * @return the parameter's value.
+     */
+    public double getRankingThreshold() {
+        return qrRankingThreshold;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Optimum optimize(final LeastSquaresProblem problem) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 43. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.